crates/matrix-sdk/src/event_cache/paginator.rs

```diff
@@ -0,0 +1,564 @@
+            );
+            assert_eq!(opts.dir, Direction::Backward, "we must receive a single back-pagination");
+
+            self.room_ready.notified().await;
+
+            let event_factory = EventFactory::new().room(*ROOM_ID).sender(*USER_ID);
+
+            let chunk = (20..30)
+                .rev()
+                .map(|i| {
+                    TimelineEvent::new(event_factory.text_msg(format!("{i}")).into_raw_timeline())
+                })
+                .collect();
+
+            return Ok(Messages {
+                start: opts.from.unwrap().to_owned(),
+                end: None,
+                chunk,
+                state: Vec::new(),
+            });
+        }
+    }
+
+    async fn assert_invalid_state<T: std::fmt::Debug>(
+        task: impl Future<Output = Result<T, PaginatorError>>,
+        expected: PaginatorState,
+        actual: PaginatorState,
+    ) {
+        assert_let!(
+            Err(PaginatorError::InvalidPreviousState {
+                expected: real_expected,
+                actual: real_actual
+            }) = task.await
+        );
+        assert_eq!(real_expected, expected);
+        assert_eq!(real_actual, actual);
+    }
+
+    #[async_test]
+    async fn test_state() {
+        let room = Box::<DummyRoom>::default();
+
+        let paginator = Arc::new(Paginator::new(room.clone()));
+
+        let event_id = event_id!("$yoyoyo");
+
+        let mut state = paginator.state();
+
+        assert_eq!(state.get(), PaginatorState::Initial);
+        assert!(state.next().now_or_never().is_none());
+
+        // Attempting to run pagination must fail and not change the state.
+        assert_invalid_state(
+            paginator.paginate_backward(),
+            PaginatorState::Idle,
+            PaginatorState::Initial,
+        )
+        .await;
+
+        assert!(state.next().now_or_never().is_none());
+
+        // Running the initial query must work.
+        let p = paginator.clone();
+        let join_handle = spawn(async move { p.start_from(event_id).await });
+
+        assert_eq!(state.next().await, Some(PaginatorState::FetchingTargetEvent));
+        assert!(state.next().now_or_never().is_none());
+
+        // The query is pending. Running other operations must fail.
+        assert_invalid_state(
+            paginator.start_from(event_id),
+            PaginatorState::Initial,
+            PaginatorState::FetchingTargetEvent,
+        )
+        .await;
+
+        assert_invalid_state(
+            paginator.paginate_backward(),
+            PaginatorState::Idle,
+            PaginatorState::FetchingTargetEvent,
+        )
+        .await;
+
+        assert!(state.next().now_or_never().is_none());
+
+        // Mark the dummy room as ready. The query may now terminate.
+        room.mark_ready();
+
+        // After fetching the initial event data, the paginator switches to `Idle`.
+        assert_eq!(state.next().await, Some(PaginatorState::Idle));
```

Comment:
I would prefer a separate test for the content of the response. All this state stuff is very important but also very distracting from the "normal" case of "when I ask for event, I get events".